Add support for 24-bit / full RGB color

dandavison opened the pull request:
Hi, thanks very much for this project. This PR adds support for 24-bit / full RGB ANSI color sequences. I have time available to do further work on this branch if that would help.

## Changes

1. Add support for 24-bit ANSI color sequences
2. Add an option allowing the user to request the colors as overlays instead of text properties.
3. ~~Make it possible for the user to specify that `'face` rather than `'font-lock-face` is used, even when `font-lock-mode` is active.~~

<details>

I'm using this to add support in magit for handling ANSI colors when the git output is redirected through a diff renderer such as [delta](https://github.com/dandavison/delta). To motivate the new options regarding overlays ~~and faces~~, it might be helpful to show the function that I am currently using in magit to apply `xterm-color` to the output received from git (via `delta`):

```emacs-lisp
(defun magit-diff-convert-ansi-colors--xterm-color (beg end)
  (require 'xterm-color)
  ;; save-excursion does not work here because xterm-color uses delete-and-extract-region
  (let ((pos (point)))
    (save-restriction
      (narrow-to-region beg end)
      (let ((inhibit-read-only t)
            (buffer-read-only nil))
        (xterm-color-colorize-buffer 'use-overlays)))
    (goto-char pos)))
```

</details>

## How do we know this is correct?

1. I've added a 24-bit color demo to `xterm-color-test`:
    <table><tr><td><img width=500px src="https://user-images.githubusercontent.com/52205/79007717-bcd25d00-7b29-11ea-9546-5699ffb6ca8f.png" alt="image" /></td></tr></table>

    It looks plausibly correct. Furthermore, the 24-bit colors in the demo are the same as in this [shell-script](https://github.com/gnachman/iTerm2/blob/master/tests/24-bit-color.sh). When the shell script is run in a terminal emulator (iTerm2), it appears the same:
    <table><tr><td><img width=500px src="https://user-images.githubusercontent.com/52205/79008397-52222100-7b2b-11ea-8d77-da87607bb2fd.png" alt="image" /></td></tr></table>

2. **What about the rest of the output in `xtem-color-test`**?<br>
   Here are screenshots of the test buffers on the two branches:
   <table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79022251-7ee63080-7b4b-11ea-9385-add419b0d855.png" alt="image" /></td></tr></table>
   <table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79022307-a0dfb300-7b4b-11ea-8fd9-69ccc06f5dd6.png" alt="image" /></td></tr></table>
   <table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79022383-c8368000-7b4b-11ea-9c22-641b14438beb.png" alt="image" /></td></tr></table>
   <table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79022451-ee5c2000-7b4b-11ea-93fe-b9eac1d87c4b.png" alt="image" /></td></tr></table>

3. **When `xterm-color` encounters a 24-bit ANSI sequence, the face cache key is > 32bits. Does that work in Emacs on a 32-bit architecture?**<br>
   I wasn't sure what to expect the answer to be here. What I have done is create an i686 ubuntu VM (`Linux ubuntu-xenial 4.4.0-177-generic #207-Ubuntu SMP Mon Mar 16 01:15:50 UTC 2020 i686 i686 i686 GNU/Linux`). In that VM Emacs 24.5.1 has `most-positive-fixnum` equal to 536870911 (versus 2305843009213693951 on my laptop). There were no errors when running `xterm-color-test` in Emacs 24.5.1 in that VM.

4. **Any issues byte compiling?**
   `xterm-color.el` byte-compiled with no errors and appears to work correctly when using the compiled file.

5. Finally, here's a screenshot of the new code handling some more complex patterns of ANSI sequences involving foreground and background colors (this is [delta](https://github.com/dandavison/delta) in magit, using xterm-color with this PR):
    <table><tr><td><img width=300px src="https://user-images.githubusercontent.com/52205/79015227-60c40480-7b3a-11ea-8eed-de774b991ff0.png" alt="image" /></td></tr></table>


## Has it impacted 256-color performance?

  Here's a shell script we can use to investigate that:
  ```emacs-lisp
  #!/usr/bin/env emacs --script
  (let ((package-load-list '((xterm-color t)))) (package-initialize))
  (require 'xterm-color)
  (let ((n (string-to-number (getenv "XTERM_COLOR_TEST_N_ITERATIONS"))))
    (dolist (i (number-sequence 0 (1- n)))
      (xterm-color-test)))
  ```

  The experiment I'm doing is looking at the impact of the proposed branch on the existing 256-color processing, so I'm comparing `master` (4b21b619841c93c4700039a93eb1881beee9248c) against this PR, but with the change to `xterm-color-test` reverted.

  The results show a very small decrease in speed: it looks like each call to `xterm-color-test` takes approximately 3ms longer.

EDIT: these timings are without byte compiling; see comment below for byte-compiled times.

  ```
   xterm-color  (master) for i in 1 2 3 4 5; do XTERM_COLOR_TEST_N_ITERATIONS=100 time bash -c "./time-xterm-color-test  2> /dev/null" ; done
         10.42 real        10.15 user         0.05 sys
         10.32 real        10.07 user         0.05 sys
         10.12 real         9.87 user         0.05 sys
         10.12 real         9.87 user         0.04 sys
         10.11 real         9.88 user         0.04 sys
   xterm-color  (master) git checkout -q 24-bit-color--with-no-change-to-xterm-color-test
   xterm-color  (24-bit-color--with-no-change-to-xterm-color-test) for i in 1 2 3 4 5; do XTERM_COLOR_TEST_N_ITERATIONS=100 time bash -c "./time-xterm-color-test  2> /dev/null" ; done
         10.71 real        10.45 user         0.05 sys
         10.46 real        10.22 user         0.04 sys
         10.45 real        10.22 user         0.05 sys
         10.42 real        10.21 user         0.04 sys
         10.40 real        10.15 user         0.05 sys
  ```

dandavison commented:
Regarding #10, I did have signed FSF papers in the past, and I am happy to check up on my status there and renew if necessary.

atomontage commented:
Thanks for the thorough work Dan! I'll try and carve out some time for review and merge.
Out of curiosity, how does using face rather than font-lock-face with font-locking active, come into play in practice?

dandavison commented:
> I'll try and carve out some time for review and merge.

Thanks!

> how does using face rather than font-lock-face with font-locking active, come into play in practice?

Ah, good call, I think you identified a part of the PR that needed revisiting. When using xterm-color to display colors in magit, I was originally having trouble with the faces being volatile: getting wiped out by font-lock in the magit buffer. However, your comment prompted me to double-check that I could reproduce that on the current version, and I cannot. So I think it was a mis-step on my part to have concluded that I should use `'face` in conjunction with magit.

So, what I've done is push d858d1952adbfaa66f9fa2d9da77479e6209d785, which reverts the option to specify the face property. It uses a buffer-local variable to prevent us from choosing the face dynamically, twice.  I suspect that that latest commit is more similar to what we're going to end up deciding on, and I think it's easier for us to discuss with the smaller diff against master.

I have also force-pushed the branch, because I realized that the commit message on the main commit was very inaccurate, due to a subsequent simplification. (I won't need to force push again).

dandavison commented on xterm-color.el:
```diff
@@ -831,24 +920,26 @@ effect when called from a buffer that does not have a cache."
 (defun xterm-color-test ()
   "Create/display and render a new buffer that contains ANSI control sequences."
```
The diff in this function is noisy due to an indentation change. Telling Github to hide whitespace, it looks like this:

<table><tr><td><img width=400px src="https://user-images.githubusercontent.com/52205/79035462-fb4d3380-7b8c-11ea-896a-cbf32830eecb.png" alt="image" /></td></tr></table>

dandavison commented:
I realized that that the benchmarks above were made without byte-compiling. Here are timings of the compiled code (256 color codes only). I think the conclusion is the same -- no difference that users will be able to notice: each call to `xterm-color-test` is less than 1ms slower on this branch.

```
 xterm-color  (master) hyperfine 'bash -c "XTERM_COLOR_TEST_N_ITERATIONS=100 ./time-xterm-color-test  2> /dev/null"'
Benchmark #1: bash -c "XTERM_COLOR_TEST_N_ITERATIONS=100 ./time-xterm-color-test  2> /dev/null"
  Time (mean ± σ):      3.501 s ±  0.095 s    [User: 3.336 s, System: 0.040 s]
  Range (min … max):    3.352 s …  3.633 s    10 runs

 xterm-color  (24-bit-color--with-no-change-to-xterm-color-test) hyperfine 'bash -c "XTERM_COLOR_TEST_N_ITERATIONS=100 ./time-xterm-color-test  2> /dev/null"'
Benchmark #1: bash -c "XTERM_COLOR_TEST_N_ITERATIONS=100 ./time-xterm-color-test  2> /dev/null"
  Time (mean ± σ):      3.556 s ±  0.075 s    [User: 3.382 s, System: 0.041 s]
  Range (min … max):    3.463 s …  3.671 s    10 runs
```

atomontage commented on xterm-color.el:
```diff
@@ -497,11 +532,18 @@ in LIFO order."
-                                                 (or xterm-color--current-fg 0))
+          (face-cache-get ()   `(gethash (let ((fg (or xterm-color--current-fg 0))
+                                               (bg (or xterm-color--current-bg 0)))
+                                           (if (has? +24-bit+)
+                                               (logior (ash xterm-color--attributes 48)
+                                                       (ash bg 24)
+                                                       fg)
```
This is broken on 32bit Emacs:

```
ELISP> (ash 1 48)
0 (#o0, #x0, ?\C-@)
```
To fix this, you would use a different scheme to store the color/attribute combination in the hash table. Now, the reason I added the cache to begin with was to avoid consing, since this is the hot path. This implies that a simple scheme of (cons attributes (cons fg bg)) is not ideal. 

I would store fg and bg (48 bits) in a float that can store up to (expt 2 53) without loss of precision. You can't
use bit operations but you can add and multiply:

```
ELISP> (+ (* #xAABBCC (float (expt 2 24)))                                                                                                                               
          #x112233)
187723559281203.0
ELISP> (floor *)
187723559281203 (#o5253571404221063, #xaabbcc112233)
```
You can't store attributes in the same float too though, since you'll be exceeding range by 2 bits (you don't need to store the truecolor bit if you use a separate hash table). You could try a multi-level hash table. You should test only at load time if Emacs is 32bit or 64bit (e.g. (> most-positive-fixnum (expt 2 56.0))

Now this is starting to get very messy as you have two paths: 64bit Emacs with a single face cache, 32bit Emacs with two face caches and the xterm-color--with-ANSI-macro-helpers macro is already unwieldy. So in the interests of keeping the code simple, I say don't bother and don't add support for truecolor on 32bit Emacs. If you still want to give it a shot go ahead, but I won't merge it if I think that it overcomplicates  the rest of the code. It's simply not worth it.

atomontage commented on xterm-color.el:
```diff
@@ -516,27 +558,33 @@ in LIFO order."
                                                        (<= 8 xterm-color--current-fg 15)))
                                               (progn (face! :weight 'bold)
                                                      (face! :foreground
-                                                            (xterm-color-256 (if (<= 8 xterm-color--current-fg)
-                                                                                 (- xterm-color--current-fg 8)
-                                                                               xterm-color--current-fg))))
+                                                            (if (has? +24-bit+)
+                                                                (xterm-color-24-bit xterm-color--current-fg)
+                                                              (xterm-color-256 (if (<= 8 xterm-color--current-fg)
+                                                                                   (- xterm-color--current-fg 8)
+                                                                                 xterm-color--current-fg)))))
                                             (face! :foreground
-                                                   (xterm-color-256
-                                                    (if (and (<= xterm-color--current-fg 7)
-                                                             (has? +bright+))
-                                                        (+ xterm-color--current-fg 8)
-                                                      xterm-color--current-fg))))
+                                                   (if (has? +24-bit+)
+                                                       (xterm-color-24-bit xterm-color--current-fg)
+                                                     (xterm-color-256
+                                                      (if (and (<= xterm-color--current-fg 7)
+                                                               (has? +bright+))
+                                                          (+ xterm-color--current-fg 8)
+                                                        xterm-color--current-fg)))))
                                         (when (and xterm-color-use-bold-for-bright
                                                    (has? +bright+))
                                           (face! :weight 'bold)))
                                       (when xterm-color--current-bg
-                                        (face! :background (xterm-color-256 xterm-color--current-bg)))
+                                        (face! :background (if (has? +24-bit+)
+                                                               (xterm-color-24-bit xterm-color--current-bg)
+                                                             (xterm-color-256 xterm-color--current-bg))))
                                       (setf (face-cache-get) plistf))))
           (maybe-fontify ()    '(when xterm-color--char-list
                                   (let ((s (concat (nreverse xterm-color--char-list))))
                                     (when (color?)
                                       (add-text-properties
                                        0 (length s)
-                                       (list 'xterm-color t (if font-lock-mode 'font-lock-face 'face) (make-face))
+                                       (list 'xterm-color t xterm-color--face-property (make-face))
```
Revert to previous version.

atomontage commented on xterm-color.el:
```diff
@@ -497,11 +532,18 @@ in LIFO order."
+                                                     fg)))
                                          xterm-color--face-cache))
           (face! (k v)         `(setq plistf (plist-put plistf ,k ,v)))
+          (xterm-color-24-bit (color) `(format "#%06x" ,color))
```
Rename to truecolor. It's inside a macrolet, no reason to add the xterm-color prefix.
Someone asked me to make xterm-color-256 part of the external API which is why it's not in the macrolet too.

atomontage commented on xterm-color.el:
```diff
@@ -656,15 +704,25 @@ This can be inserted into `comint-preoutput-filter-functions'."
-  "Apply `xterm-color-filter' to current buffer, and replace its contents."
-  (interactive)
-  (let ((read-only-p buffer-read-only))
+(cl-defun xterm-color-colorize-buffer (&optional use-overlays?)
+  "Apply `xterm-color-filter' to current buffer, and replace its contents.
+
+The colors will be applied using 'font-lock-face, unless
+font-lock-mode is inactive, in which case 'face will be used.
+
+If USE-OVERLAYS? is non-nil then the colors will be applied to
+the buffer using overlays instead of text properties. A C-u
+prefix arg causes overlays to be used."
```
Rename use-overlays? to use-overlays

atomontage commented:
I pushed a few changes today to master, mostly documentation updates and a parsing improvement (don't signal an error that can desync the state machine). You should be able to easily resolve any conflicts.

dandavison commented:
Thanks for the very helpful review @atomontage. I've merged master and I believe I've addressed all the points.

I agree with not complicating the code for machines with 32 bit integers. What I've done in  85a3e240bd is disable caching of faces when the current fg/bg is a full RGB ANSI sequence and the integer width is insufficient. Do you think that's reasonable?

Would you like to make an executive decision on which of, and how, the terms "full RGB", "truecolor" and "24-bit" should be used? The code now uses `"truecolor"` as you requested. However there are a few places that are bothering me:

```emacs-lisp
;; SGR attributes 38 and 48 are supported in both their 8-bit (256 color) and 24-bit
;; (full RGB) variants.
```

```emacs-lisp
(defsubst xterm-color--skip-24 (SGR-list)
```

```emacs-lisp
                  (eq 2 (cl-second SGR-list)))          ; 24-bit FG color
             'xterm-color--skip-24)
            (set-a! +truecolor+)
            (set-truecolor! (cddr SGR-list) xterm-color--current-fg))
    (:match ((and (eq 48 (cl-first SGR-list))
                  (eq 2 (cl-second SGR-list)))          ; 24-bit BG color
```

```emacs-lisp
   ;; 24-bit colors
   (insert "\n")
   (insert "*  24-bit color ramps\n\n")
```

```emacs-lisp
;; The face caching scheme requires an integer width of at least 56 bits to cache faces derived
;; from truecolor (i.e. full RGB / 24-bit) ANSI sequences. Caching is therefore disabled on e.g.
;; machines with 32-bit integers.
(defvar xterm-color--cache-truecolor-faces (>= most-positive-fixnum (1- (expt 2 56.0))))
```

dandavison commented on xterm-color.el:
```diff
@@ -505,11 +546,19 @@ in LIFO order."
+                                                         fg))
+                                             (logior (ash xterm-color--attributes 16)
+                                                     (ash bg 8)
+                                                     fg)))
```
Sorry, this isn't going to behave correctly on 32bit either. I'm going to push a further commit.

dandavison commented:
At the latest commit, 4baa93ba0b2dadbb43e51e2c29e4a08ca1436f8a, 24-bit color support is disabled when the integer width is insufficient. It's disabled in the sense that the state machine will not recognize those ANSI sequences as 24-bit, so the behavior is as on master.

atomontage commented:
I would use truecolor everywhere, and truecolor (24-bit) in comments next to relevant code to make the structure ultra obvious.

dandavison commented:
OK, 760518e22e543db8e5916477427182c9bb989c6c makes terminology more consistent. It also changes the library description in the first line of the file to 

```emacs-lisp
;;; xterm-color.el --- ANSI, XTERM 256, and Truecolor color support -*- lexical-binding: t -*-
```

Of course please feel free to push to this branch to edit any of this, e.g. capitalization.

atomontage commented:
I will do another review pass soon and merge. There's a few spots I'll fix up myself later.
Thanks again.

dandavison commented:
Squashed. I added one line to the README.

dandavison commented on xterm-color.el:
```diff
@@ -310,6 +314,24 @@ inverse-color, frame, overline SGR state machine bits.")
        (cl-return-from xterm-color--string-properties (nreverse result))))))
 
+(defun xterm-color--convert-text-properties-to-overlays (beg end)
+  "Delete face text properties between BEG and END, replacing with equivalent overlays."
```
My current motivation for adding this function is for usage with Magit. When displaying a diff, Magit currently translates ANSI color sequences received from git into overlays (using the `ansi-color` library that is in Emacs). I am making it possible for Magit to use `xterm-color` instead of `ansi-color`, but I don't want to conflate that change with a change from overlays to text properties in the Magit diff buffer. Indeed, empirically, I see that if I set `xterm-color` to create text properties instead of overlays then the fontification in the Magit diff buffer  no longer works, so I suspect that Magit's choice of overlays may be deliberate. I think that other users of `xterm-color` will also need to be able to make this choice.

atomontage commented:
Did some extra work and cleanup, improved performance and parsing reliability and the ANSI-macro-helpers macro should be a lot easier to read now. There are a few more things I want to commit and then I'll release 2.0.

dandavison commented:
Great. That all looks much cleaner, in particular `make-color-fg` and `make-color-bg`.

atomontage commented:
I pushed more fixes. Basically splitting the truecolor representation out of the color itself and having a separate attribute/flag to keep it doesn't work. It seems like it works and the tests that exist seem to work but it's fundamentally broken.

Example:

`printf "\x1b[38;2;100;250;250;42mHELLO\x1b[m\n"`

Here, the truecolor "flag" stays on after the 38;2;100;250;250 sequence
is parsed, since it's not explicitly being turned off. End result being that the
next sequence is totally mishandled. Treating the color as a color and not
as extra state solves this issue

I should have caught this earlier since I've been bitten by a similar issue when 
the AIXTERM pull request came in and I wasn't paying attention. That goes to show
you that things that may appear simple, can hide unexpected complexity. You should
pause and look at things again and again, whenever you're dealing with non-trivial
state machines.

dandavison commented:
I see, sorry about that, thanks for working on this so carefully. What do you think, in principle, about adding some `ert` tests, and a CI build?

atomontage commented:
No worries, these are mostly notes to myself. Thanks again, if it weren't for your pull request waking me up from my slumber, there wouldn't be truecolor support!

I will add some more tests but they won't be exhaustive. That's a lot of work as you'll need to test against another state machine that you trust to be correct, and I simply don't have the time. Setting all of this up is outside of the scope of what I envisioned for this library. Maybe one day.

dandavison commented:
OK, I have another repo with an `ert` test suite, a Makefile and a .travis.yml config that are already set up and working. I sent you a collaborator invitation (it's private currently) in case you're interested in copying that setup over; I'd be happy to help.

atomontage commented:
And of course it's still not correct, my new packing scheme works to keep everything in one integer but not in the cache as there's overlap since one bit of information (is the color present or not, the variable is set to nil outside the cache if it's not, but it's zeroed in the cache which overlaps) is lost! Even my old 256 color caching scheme was not correct.

I *think* I have it now, but I'm also in a haze as I'm about to go to sleep. I'll give it another shot tomorrow.
